### ISSUE TITLE
Engine returns assembly-level test-suite event twice

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit-console#603]
I noted this phenomenon in working on the TestCentric GUI. The `<test-suite />` event is sent twice for each assembly. I have a test that verifies this problem but it's based on my unmerged PR #596 so I can't submit it just yet.

Currently, this is not tested - we only check that an assembly-level event is received, not that it's only received once. Clearly, this behavior didn't affect any other runners. I only noticed it because I was counting down the number of assemblies completed.

I'm reporting this against the engine because that's where I encountered it. It may, of course, turn out to be the framework that sends the event twice.

**Comment by CharliePoole** [comment]
After investigation, I think I have found the source of the problem. The framework contributes to it in the way it returns results but it's really how the NUnit3FrameworkDriver interprets them that is the source of the problem.

All the actions performed by the driver produce a result. In the case of the Run action, intermediate progress events are also produced - i.e. test listener events.

In NUnit V2, the framework sent events separately from returned results. With NUnit 3, we started to use the ICallbackEventHandler interface, which only provides a single method for both purposes. Ideally, the string value we pass should have been encoded in some way that allowed easy separation of intermediate events from results. Unfortunately, that's not how I did it. Even if some future version of the framework were to wrap the final result in (e.g.) a `<final-result>` element, the driver would still have to deal with older versons of the frameowork. (However, I do think that might be a good idea!)

So what happens is this:

1. The last test suite to be run is the one that represents the assembly. An event is sent and propagated by the engine.

2. The final result is then returned using the same callback mechanism. Code in the driver sets the result and then propagates the event. Most runners we have had up to now don't care that they were notified twice. As often happens, the GUI is more demanding and gets confused by the second notification.

We need to return to general logic of NUnit V2, while continuing to use the `ICallbackEventHandler` interface. That is, we need to look at each callback and decide whether it is the final result or an interim event. Currently, we can only do that by use of the content of the `test-suite` element.

The key difference between the two is that the `test-suite` element for the final result contains some additional child elements (e.g. `environment`) and __may__ also contain nested `test-suites`. It won't have any nested suites if, for example, the file could not be found or was otherwise unusable.

I'm taking a two-pronged approach here.

1. In the engine, set up some heuristics based on nested elements to distinguish the final result from true progress events. Only save the final result. Only propagate true events.

2. Consider wrapping the final result XML with a new element in the framework. If we pre-define what that will be, then we have the engine use it immediately. Otherwise a second engine change will be needed.
